lib/Driver/ToolChain.cpp

```diff
@@ -62,6 +62,27 @@ ToolChain::JobContext::getTemporaryFilePath(const llvm::Twine &name,
   return C.getArgs().MakeArgString(buffer.str());
 }
 
+ToolChain::ResponseFileInfo ToolChain::getResponseFileInfoIfNeeded(
+    const Compilation &C, const char *executablePath,
+    const ToolChain::InvocationInfo &invocationInfo,
+    const ToolChain::JobContext &context) const {
+  const bool forceResponseFiles =
+      C.getArgs().hasArg(options::OPT_driver_force_response_files);
+  assert((invocationInfo.allowsResponseFiles || !forceResponseFiles) &&
+         "Cannot force response file if platform does not allow it");
+
+  if (forceResponseFiles || (invocationInfo.allowsResponseFiles &&
+                             !llvm::sys::commandLineFitsWithinSystemLimits(
+                                 executablePath, invocationInfo.Arguments))) {
+    const char *responseFilePath =
+        context.getTemporaryFilePath("arguments", "resp");
+    const char *responseFileArg =
+        C.getArgs().MakeArgString(Twine("@") + responseFilePath);
+    return {responseFilePath, responseFileArg};
```

Comment:
I suspect does not work: copies pointers then the stack frame is popped. Suggest std::string instead.

Reply:
This does work, the same as it did before this change (note that nothing I've changed here affects the lifetimes of those strings, it just extracted the code to a new function).

The `MakeArgString` function returns a heap-allocated string whose lifetime is tied to that of the `Compilation`, which is what makes all this work because it guarantees that the strings stay alive as long as the jobs do. Likewise, `context.getTemporaryFilePath` also calls `MakeArgString` under the hood.

Reply:
Okey-doke!